port/unix/omrsignal.c

```diff
@@ -1227,6 +1253,30 @@ static void
 mainASynchSignalHandler(int signal, siginfo_t *sigInfo, void *contextInfo)
 #endif /* defined(S390) && defined(LINUX) */
 {
+	do {
+		uintptr_t tail = signalPidTails[signal];
+		uintptr_t nextTail = (tail + 1) % NUM_SIGNAL_PIDS;
+		uintptr_t head = signalPidHeads[signal];
+		/* Check if there is room. */
+		if (nextTail != head) {
+			if (sizeof(sigInfo->si_pid) == sizeof(uint32_t)) {
+				if ((uint32_t)-1 == compareAndSwapU32((uint32_t *)&signalPids[signal][tail], (uint32_t)-1, sigInfo->si_pid)) {
+					/* The pid is written before the tail is incremented. */
+					signalPidTails[signal] = nextTail;
+					break;
+				}
+			} else {
+				if ((uint64_t)-1 == compareAndSwapUDATA((uint64_t *)&signalPids[signal][tail], (uint64_t)-1, sigInfo->si_pid)) {
+					/* The pid is written before the tail is incremented. */
+					signalPidTails[signal] = nextTail;
+					break;
+				}
+			}
```

Comment:
`UDATA` and `uint64_t` aren't necessarily the same size without a check; suggest inverting this:
```c
			if (sizeof(sigInfo->si_pid) == sizeof(uintptr_t)) {
				if ((uintptr_t)-1 == compareAndSwapUDATA((uintptr_t *)&signalPids[signal][tail], (uintptr_t)-1, sigInfo->si_pid)) {
					/* The pid is written before the tail is incremented. */
					signalPidTails[signal] = nextTail;
					break;
				}
			} else {
				if ((uint32_t)-1 == compareAndSwapU32((uint32_t *)&signalPids[signal][tail], (uint32_t)-1, sigInfo->si_pid)) {
					/* The pid is written before the tail is incremented. */
					signalPidTails[signal] = nextTail;
					break;
				}
			}
```
The `if` on line 1261 could also be inverted to reduce the indentation here.